hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/CassAlertsServiceImpl.java

```diff
@@ -0,0 +1,849 @@
+
+        wakeUpTimer = new Timer("CassAlertsServiceImpl-Timer");
+
+        delay = new Integer(AlertProperties.getProperty(ENGINE_DELAY, "1000"));
+        period = new Integer(AlertProperties.getProperty(ENGINE_PERIOD, "2000"));
+    }
+
+    @PostConstruct
+    public void initServices() {
+        try {
+            if (this.keyspace == null) {
+                this.keyspace = AlertProperties.getProperty(CASSANDRA_KEYSPACE, "hawkular_alerts");
+            }
+
+            if (session == null) {
+                session = CassCluster.getSession();
+            }
+
+            initPreparedStatements();
+
+            GsonBuilder gsonBuilder = new GsonBuilder();
+            gsonBuilder.registerTypeHierarchyAdapter(ConditionEval.class, new GsonAdapter<ConditionEval>());
+            gson = gsonBuilder.create();
+
+            reload();
+        } catch (Throwable t) {
+            msgLog.errorCannotInitializeAlertsService(t.getMessage());
+        }
+    }
+
+    private void initPreparedStatements() throws Exception {
+        if (insertAlert == null) {
+            insertAlert = session.prepare("INSERT INTO " + keyspace + ".alerts " +
+                    "(alertId, payload) VALUES (?, ?) ");
+        }
+        if (insertAlertTrigger == null) {
+            insertAlertTrigger = session.prepare("INSERT INTO " + keyspace + ".alerts_triggers " +
+                    "(alertId, triggerId) VALUES (?, ?) ");
+        }
+        if (insertAlertCtime == null) {
+            insertAlertCtime = session.prepare("INSERT INTO " + keyspace + ".alerts_ctimes " +
+                    "(alertId, ctime) VALUES (?, ?) ");
+
+        }
+        if (insertAlertStatus == null) {
+            insertAlertStatus = session.prepare("INSERT INTO " + keyspace + ".alerts_statuses " +
+                    "(alertId, status) VALUES (?, ?) ");
+        }
+        if (updateAlert == null) {
+            updateAlert = session.prepare("UPDATE " + keyspace + ".alerts " +
+                    "SET payload = ? WHERE alertId = ? ");
+        }
+        if (selectAlertStatus == null) {
+            selectAlertStatus = session.prepare("SELECT alertId, status FROM " + keyspace + ".alerts_statuses " +
+                    "WHERE alertId = ? ALLOW FILTERING ");
+        }
+        if (deleteAlertStatus == null) {
+            deleteAlertStatus = session.prepare("DELETE FROM " + keyspace + ".alerts_statuses " +
+                    "WHERE alertId = ? AND status = ?");
+        }
+    }
+
+    public void addAlerts(Collection<Alert> alerts) throws Exception {
+        if (alerts == null) {
+            throw new IllegalArgumentException("Alerts must be not null");
+        }
+        if (session == null) {
+            throw new RuntimeException("Cassandra session is null");
+        }
+        if (insertAlert == null) {
+            throw new RuntimeException("insertAlert PreparedStatement is null");
+        }
+        try {
+            for (Alert a : alerts) {
```

Comment:
+1
Yes, parallel async insert is a good point.
But the business method would be sync, I mean, I think that propragation of the async cappabilities should be (at the moment) in the implementation detail.